backend/ingestion/extract_census_data.py

```diff
@@ -1 +1,147 @@
- 
\ No newline at end of file
+import os
+import sys
+import csv
+import logging
+import requests
+from typing import List, Dict
+from dotenv import load_dotenv
+
+########################################################################################
+# SETUP
+########################################################################################
+
+# Enviornment variables
+load_dotenv()
+CENSUS_API_KEY = os.getenv("CENSUS_API_KEY")
+
+# Logging configuration
+logging.basicConfig(
+    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
+)
+
+# Mappings
+variable_ids = {"B01001_001E": "total_population"}
+city_fips = {
+    "nyc": {"state_fips": "36", "county_fips": ["061", "047", "081", "005", "085"]},
+    "chicago": {"state_fips": "17", "county_fips": ["031"]},
+    "portland": {"state_fips": "41", "county_fips": ["051"]},
+}
+
+########################################################################################
+# HELPER FUNCTIONS
+########################################################################################
+
+
+def valid_command_line_arg(supported_cities: List[str]) -> str:
+    """
+    Validates command line use and extracts argument.
+
+    Inputs:
+        available_cities (List[str]): List of cities that are valid inputs.
+
+    Returns:
+      A string referring to the city to pull data from.
+    """
+    # Check syntax
+    if len(sys.argv) != 2:
+        sys.exit("Retype command as 'python3 extract_census_data.py <city_name>'")
+    # Check city
+    city = sys.argv[1]
+    if city not in supported_cities:
+        sys.exit(
+            "Unsupported city." + f" Available options: {', '.join(supported_cities)}"
+        )
+    return city
+
+
+def get_census_data(
+    variable_ids: Dict[str, str], state_code: str, county_code: str
+) -> List[Dict]:
+    """
+    Fetches data from US Census API.
+
+    Inputs:
+        variable_ids (dict): Unique variable IDs and their names
+        state_code (str): State-level FIPS code
+        county_code (str): County-level FIPS code
+
+    Returns:
+        A list of dictionaries with function inputs as key, value pairs.
+    """
+    # Request parameters
+    url = "https://api.census.gov/data/2022/acs/acs5"
+    params = {
+        "get": ",".join(variable_ids.keys()),
+        "for": "block group:*",
+        "in": f"state:{state_code}+county:{county_code}",
+        "key": CENSUS_API_KEY,
+    }
+    # API call
+    response = requests.get(url, params=params)
+    if response.status_code == 200:
+        data = response.json()
+        headers = data[0]
+        rows = data[1:]
+        formatted_data = [dict(zip(headers, row)) for row in rows]
+        return formatted_data
+    else:
+        logging.error(f"Failed to retrieve data. Status code: {response.status_code}")
+        return []
+
```

Comment:
Note to self: Curious about best practices here that we should explore.
Maybe with some error codes we might want to retry, and with others just return an empty list or personalize the error handling depending on the status code